hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java

```diff
@@ -494,20 +459,87 @@ private void handleRetry(Object cause) {
                 operationService.asyncExecutor.execute(this);
             } else {
-                nodeEngine.getExecutionService().schedule(ASYNC_EXECUTOR, this, tryPauseMillis, MILLISECONDS);
+                if (invokeCount < MAX_FAST_INVOCATION_COUNT) {
+                    // fast retry for the first few invocations
+                    operationService.asyncExecutor.execute(this);
+                } else {
+                    nodeEngine.getExecutionService().schedule(ASYNC_EXECUTOR, this, tryPauseMillis, MILLISECONDS);
+                }
             }
         }
     }
 
-    boolean checkBackupTimeout(long timeoutMillis) {
+    /**
+     * Checks if this Invocation has received a heart-beat in time.
+     *
+     * If the response is already set, or if a heart-beat has been received in time, then the false is returned.
+     *
+     * If no heart-beat has been received, then the future.set is called with HEARTBEAT_TIMEOUT
+     * and true is returned.
+     *
+     * gets called from the monitor-thread
+     *
+     * @return true if there is a timeout detected, false otherwise.
+     */
+    boolean detectAndHandleTimeout(long heartbeatTimeoutMillis) {
+        HeartbeatTimeout heartbeatTimeout = detectTimeout(heartbeatTimeoutMillis);
+
+        if (heartbeatTimeout == YES) {
+            future.complete(HEARTBEAT_TIMEOUT);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    HeartbeatTimeout detectTimeout(long heartbeatTimeoutMillis) {
+        if (pendingResponse != VOID) {
+            // if there is a response, then we won't timeout.
+            return NO__RESPONSE_AVAILABLE;
+        }
+
+        long callTimeoutMillis = op.getCallTimeout();
+        if (callTimeoutMillis <= 0 || callTimeoutMillis == Long.MAX_VALUE) {
+            return NO__CALL_TIMEOUT_DISABLED;
+        }
+
+        // a call is always allowed to execute as long as its own call timeout.
+        long deadlineMillis = op.getInvocationTime() + callTimeoutMillis;
+        if (deadlineMillis > nodeEngine.getClusterService().getClusterClock().getClusterTime()) {
+            return NO__CALL_TIMEOUT_NOT_EXPIRED;
+        }
+
+        // on top of its own call timeout, it is allowed to execute until there is a heart beat timeout.
+        // so if the callTimeout = 5m, and the heartbeatTimeout = 1m, then it allowed to execute for
+        // at least 6 minutes before it is timing out.
+        long lastHeartbeatMillis = this.lastHeartbeatMillis;
+        long heartBeatExpirationTimeMillis = lastHeartbeatMillis == 0
+                ? op.getInvocationTime() + callTimeoutMillis + heartbeatTimeoutMillis
+                : lastHeartbeatMillis + heartbeatTimeoutMillis;
+
+        if (heartBeatExpirationTimeMillis >  Clock.currentTimeMillis()) {
+            return NO__HEARTBEAT_TIMEOUT_NOT_EXPIRED;
+        }
+
+        return YES;
+    }
+
+    enum HeartbeatTimeout {
```

Comment:
The problem is then the testing becomes very problematic because you don't know why it has or has not timed out.